actionpack/lib/action_dispatch/journey/formatter.rb

```diff
@@ -70,8 +73,11 @@ def generate(name, options, path_parameters)
 
           missing_keys = missing_keys(route, parameterized_parts)
           next if missing_keys && !missing_keys.empty?
-          params = options.dup.delete_if do |key, _|
-            parameterized_parts.key?(key) || route.defaults.key?(key)
+          params = options.delete_if do |key, _|
+            # top-level params' normal behavior of generating query_params
+            # should be preserved even if the same key is also a bind_param
+            (path_params.key?(key) && !original_options.key?(key)) ||
```

Comment:
Does the performance change if we move the new conditional to the end of the expression instead of the beginning?

Reply:
Ok great question, I think moving it is a no-brainer anyway, and I'll re-run the benchmarks.

Reply:
@rafaelfranca. Answered in https://github.com/rails/rails/pull/43770#issuecomment-1523863400. Would love to see if we could get this merged before 7.1 alpha gets cut :)